ksql-functional-tests/src/main/java/io/confluent/ksql/test/tools/ExpectedRecordComparator.java

```diff
@@ -0,0 +1,157 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.test.tools;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.JsonNodeType;
+import com.fasterxml.jackson.databind.node.NumericNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+import com.google.common.collect.ImmutableMap;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import org.apache.kafka.connect.data.Struct;
+
+public final class ExpectedRecordComparator {
+
+  private static final Map<JsonNodeType, BiPredicate<Object, JsonNode>> COMPARATORS =
+      ImmutableMap.<JsonNodeType, BiPredicate<Object, JsonNode>>builder()
+          .put(JsonNodeType.OBJECT, ExpectedRecordComparator::compareStruct)
+          .put(JsonNodeType.ARRAY, ExpectedRecordComparator::compareArray)
+          .put(JsonNodeType.NUMBER, ExpectedRecordComparator::compareNumber)
+          .put(JsonNodeType.STRING, ExpectedRecordComparator::compareText)
+          .put(JsonNodeType.BOOLEAN, ExpectedRecordComparator::compareBoolean)
+          .put(JsonNodeType.NULL, ExpectedRecordComparator::compareNull)
+          .build();
+
+  private ExpectedRecordComparator() {
+  }
+
+  public static boolean matches(final Object actualValue, final JsonNode expectedValue) {
+    return comparator(expectedValue).test(actualValue, expectedValue);
+  }
+
+  private static boolean compareStruct(final Object actualValue, final JsonNode expectedValue) {
+    final ObjectNode expected = (ObjectNode) expectedValue;
+    final Function<String, Object> getter;
+    if (actualValue instanceof Struct) {
+      getter = ((Struct) actualValue)::get;
+    } else if (actualValue instanceof Map) {
+      getter = ((Map<?, ?>) actualValue)::get;
+    } else {
+      return false;
+    }
+
+    final Iterator<Entry<String, JsonNode>> fields = expected.fields();
+    while (fields.hasNext()) {
+      final Entry<String, JsonNode> field = fields.next();
+      if (!comparator(field.getValue()).test(getter.apply(field.getKey()), field.getValue())) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static boolean compareArray(final Object actualValue, final JsonNode expectedValue) {
+    final ArrayNode expected = (ArrayNode) expectedValue;
+    if (actualValue instanceof List) {
+      final List<?> actual = (List<?>) actualValue;
+      final Iterator<JsonNode> elements = expected.elements();
+
+      int i = 0;
+      while (elements.hasNext()) {
```

Comment:
Again, is this intentionally ignore elements in actualValue beyond the end of expectedValue?

Consider comparing the size before you iterate over the elements.